geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ClientUserAuths.java

```diff
@@ -90,28 +115,29 @@ public UserAuthAttributes getUserAuthAttributes(final Long userId) {
   }
 
   @VisibleForTesting
+  @TestOnly
   protected Collection<Subject> getSubjects() {
-    return Collections.unmodifiableCollection(uniqueIdVsSubject.values());
+    List<Subject> all = uniqueIdVsSubject.values().stream()
+        .flatMap(List::stream)
+        .collect(Collectors.toList());
+    return Collections.unmodifiableCollection(all);
   }
 
-  public Subject getSubject(final Long userId) {
-    return uniqueIdVsSubject.get(userId);
+  public synchronized Subject getSubject(final Long userId) {
+    CopyOnWriteArrayList<Subject> subjects = uniqueIdVsSubject.get(userId);
+    if (subjects == null || subjects.isEmpty()) {
+      return null;
+    }
+    return subjects.get(0);
   }
 
-  public void removeSubject(final Long userId) {
+  public synchronized void removeSubject(final Long userId) {
```

Comment:
The extra synchronization is not necessary on `ConcurrentMap` and `CopyOnWriteArrayList`.

Reply:
this prevents other thread doing `put/get` at the same time.

Reply:
`CopyOnWriteArrayList` offers built-in synchronization for all write access and it makes a copy of the underlying array for all write access. If you are trying to synchronize read and write access of the list with your own synchronization, it is not necessary to use `CopyOnWriteArrayList`. Because this introduces extra overhead of synchronization and copying the array on write.